pkg/logentry/stages/template.go

```diff
@@ -0,0 +1,115 @@
+package stages
+
+import (
+	"bytes"
+	"errors"
+	"reflect"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/mitchellh/mapstructure"
+	"github.com/prometheus/common/model"
+)
+
+// Config Errors
+const (
+	ErrEmptyTemplateStageConfig = "template stage config cannot be empty"
+	ErrTemplateSourceRequired   = "template source value is required"
+	ErrTemplateRequired         = "template value is required"
+)
+
+// TemplateConfig configures template value extraction
+type TemplateConfig struct {
+	Source   string `mapstructure:"source"`
+	Template string `mapstructure:"template"`
+}
+
+// validateTemplateConfig validates the templateStage config
+func validateTemplateConfig(cfg *TemplateConfig) (*template.Template, error) {
+	if cfg == nil {
+		return nil, errors.New(ErrEmptyTemplateStageConfig)
+	}
+	if cfg.Source == "" {
+		return nil, errors.New(ErrTemplateSourceRequired)
+	}
+	if cfg.Template == "" {
+		return nil, errors.New(ErrTemplateRequired)
+	}
+
+	fm := template.FuncMap{
+		"ToLower":    strings.ToLower,
+		"ToUpper":    strings.ToUpper,
+		"Replace":    strings.Replace,
+		"Trim":       strings.Trim,
+		"TrimLeft":   strings.TrimLeft,
+		"TrimRight":  strings.TrimRight,
+		"TrimPrefix": strings.TrimPrefix,
+		"TrimSuffix": strings.TrimSuffix,
+		"TrimSpace":  strings.TrimSpace,
+	}
+	return template.New("pipeline_template").Funcs(fm).Parse(cfg.Template)
+}
+
+// newTemplateStage creates a new templateStage
+func newTemplateStage(logger log.Logger, config interface{}) (*templateStage, error) {
+	cfg := &TemplateConfig{}
+	err := mapstructure.Decode(config, cfg)
+	if err != nil {
+		return nil, err
+	}
+	t, err := validateTemplateConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &templateStage{
+		cfgs:     cfg,
+		logger:   logger,
+		template: t,
+	}, nil
+}
+
+type templateData struct {
+	Value string
+}
+
+// templateStage will mutate the incoming entry and set it from extracted data
+type templateStage struct {
+	cfgs     *TemplateConfig
+	logger   log.Logger
+	template *template.Template
+}
+
+// Process implements Stage
+func (o *templateStage) Process(labels model.LabelSet, extracted map[string]interface{}, t *time.Time, entry *string) {
+	if o.cfgs == nil {
+		return
+	}
+	if v, ok := extracted[o.cfgs.Source]; ok {
+		s, err := getString(v)
+		if err != nil {
+			level.Debug(o.logger).Log("msg", "extracted template could not be converted to a string", "err", err, "type", reflect.TypeOf(v).String())
+			return
+		}
+		td := templateData{s}
+		buf := &bytes.Buffer{}
+		err = o.template.Execute(buf, td)
+		if err != nil {
+			level.Debug(o.logger).Log("msg", "failed to execute template on extracted value", "err", err, "value", v)
+			return
+		}
+		extracted[o.cfgs.Source] = buf.String()
+	} else {
+		td := templateData{}
```

Comment:
How useful is this ? A template without data, what is the use case you see ?

Reply:
The use case in mind is to set a value to a label.  I want to make a sort of fallthrough of `match` stages that each looks for a specific word in the log via regex to set a label, if none of the keywords is found I want to default that label to a value, this would allow that to happen.

Reply:
Awesome